js-pkg/client/src/provider.ts

```diff
@@ -143,21 +154,83 @@ export class YSweetProvider {
     this.awareness.on('update', this.handleAwarenessUpdate.bind(this))
     this.WebSocketPolyfill = extraOptions.WebSocketPolyfill || WebSocket
 
+    this.online = this.online.bind(this)
+    this.offline = this.offline.bind(this)
+    if (typeof window !== 'undefined') {
+      window.addEventListener('offline', this.offline)
+      window.addEventListener('online', this.online)
+    }
+
     doc.on('update', this.update.bind(this))
 
     if (extraOptions.connect !== false) {
       this.connect()
     }
   }
 
+  private offline() {
+    // When the browser indicates that we are offline, we immediately
+    // probe the connection status.
+    // This accelerates the process of discovering we are offline, but
+    // doesn't mean we entirely true the browser, since it can be wrong
+    // (e.g. in the case that the connection is over localhost).
+    this.checkSync()
+    this.setConnectionTimeout()
```

Comment:
The trouble with that is that not every message necessarily has a response, so we might encounter false positives.

That said, I have moved the call into `checkSync`, since we know that every `checkSync` call should return a response. This also should speed up offline detection if a user is interacting with the state when the connection is dropped.